fbpic/particles/spin/spin_tracker.py

```diff
@@ -0,0 +1,276 @@
+                    self.sx, self.sy, self.sz,
+                    self.ux_prev, self.uy_prev, self.uz_prev,
+                    self.ptcl.ux, self.ptcl.uy, self.ptcl.uz,
+                    self.ptcl.Ex, self.ptcl.Ey, self.ptcl.Ez,
+                    self.ptcl.Bx, self.ptcl.By, self.ptcl.Bz,
+                    self.ptcl.q, self.ptcl.m, self.ptcl.Ntot,
+                    self.dt, self.anom)
+        # CPU version
+        else:
+            if self.ptcl.ionizer is not None:
+                # Ionizable species can have a charge that depends on the
+                # macroparticle, and hence require a different function
+                push_s_ioniz_numba(self.sx, self.sy, self.sz,
+                                   self.ux_prev, self.uy_prev, self.uz_prev,
+                                   self.ptcl.ux, self.ptcl.uy, self.ptcl.uz,
+                                   self.ptcl.Ex, self.ptcl.Ey, self.ptcl.Ez,
+                                   self.ptcl.Bx, self.ptcl.By, self.ptcl.Bz,
+                                   self.ptcl.m, self.ptcl.Ntot,
+                                   self.dt, self.anom,
+                                   self.ptcl.ionizer.ionization_level)
+                # elif z_plane is not None:
+                # ... no implemented effect on spin precession for z_plane...
+            else:
+                # Standard spin pusher
+                push_s_numba(self.sx, self.sy, self.sz,
+                             self.ux_prev, self.uy_prev, self.uz_prev,
+                             self.ptcl.ux, self.ptcl.uy, self.ptcl.uz,
+                             self.ptcl.Ex, self.ptcl.Ey, self.ptcl.Ez,
+                             self.ptcl.Bx, self.ptcl.By, self.ptcl.Bz,
+                             self.ptcl.q, self.ptcl.m, self.ptcl.Ntot,
+                             self.dt, self.anom)
+
+    def generate_new_spins(self, Ntot):
+        """
+        Create new spin vectors for particles. This method
+        generates a set of spin components, where the ensemble
+        averages satisfy the user's requirement in terms of averages.
+        """
+        if self.spin_distr == 'fixed':
+            sx = np.ones(Ntot) * self.sx_m / self.sm
+            sy = np.ones(Ntot) * self.sy_m / self.sm
+            sz = np.ones(Ntot) * self.sz_m / self.sm
+        else:
+            # If the user passes a preferred spin avergae
+            if self.sx_m != 0.:
+                sx, sy, sz = make_random_spins(Ntot, self.sx_m)
+            elif self.sy_m != 0.:
+                sy, sx, sz = make_random_spins(Ntot, self.sy_m)
+            elif self.sz_m != 0.:
+                sz, sx, sy = make_random_spins(Ntot, self.sz_m)
+            else:
+                # If the user does not pass anything, all spins
+                # are randomly oriented
+                sx, sy, sz = random_point_sphere_cpu(Ntot)
+
+        return sx, sy, sz
+
+    def generate_ionized_spins_gpu(self, Ntot):
```

Comment:
Could you add a brief docstring to this function?
In particular, it might good to say this generates the spins according to the specifications of `self.spin_distr`, `self.sx_m`, `self.sy_m`, `self.sz_m`.

Reply:
Turns out this was an oopsie: I implemented the `random_point_sphere_gpu` method directly in the ionizer. The code is much cleaner now, thanks for pointing this out.